src/handlers/dapp_transaction_handler.rs

```
@@ -113,16 +116,48 @@ pub fn supply_instructions(
     }
 
     let params_hash = {
+        let wallet = Wallet::unpack(&wallet_account_info.data.borrow())?;
+
         let mut multisig_data =
             DAppMultisigData::unpack(&multisig_data_account_info.data.borrow())?;
 
+        let whitelisting_on = wallet
+            .get_balance_account(&multisig_data.account_guid_hash)?
+            .whitelist_enabled
+            == BooleanSetting::On;
+        let instruction_whitelist = if whitelisting_on {
+            [
+                wallet
+                    .dapp_book
+                    .into_iter()
+                    .filter_map(|entry| {
+                        if entry.name_hash == multisig_data.dapp.name_hash {
+                            Some(entry.address)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<Vec<Pubkey>>(),
+                vec![
+                    solana_program::system_program::id(),
+                    spl_token::id(),
+                    spl_associated_token_account::id(),
+                ],
+            ]
+            .concat()
+        } else {
+            vec![]
+        };
         for index in starting_index..starting_index + instructions.len().as_u8() {
-            multisig_data.add_instruction(
-                index,
-                &instructions
-                    .get(usize::from(index - starting_index))
-                    .unwrap(),
-            )?;
+            let instruction = &instructions
+                .get(usize::from(index - starting_index))
+                .unwrap();
+            if whitelisting_on {
+                if !instruction_whitelist.contains(&instruction.program_id) {
+                    return Err(WalletError::DAppInstructionForbidden.into());
```

We need to think how to finalize the DAPP transaction in this case when this fails and recover any rent exemption from the op and data accounts. Currently I think there is a hole on the client side since we mark the multisig op as failed so I don;t think we would ever finalize the DAPP transaction and thus would not get the rent exemptions back and would possible leak any allocated nonce accounts. I guess it we didn't mark it as failed on client side it would eventually expire and the rent/nonce accounts could be recovered, but we also want to report the error as soon as possible to the DAPP - we might need a failed status that is still checked for expiration so we can eventually recover stuff.

---

yup, you're right, we should have a way to recover this ... on the other hand, if we do the whitelisting check in our backend (which we should anyway so we can give immediate feedback instead of forcing them to try to initiate something which will just fail), this will be a pretty rare case, and maybe we can just ignore it

---

yeah, doing the whitelisting check at init time in the backend makes alot of sense and should prevent us from really ever hitting this in the program.